### PR TITLE
[8.x] Add job middleware to prevent overlapping jobs

### PR DESCRIPTION
**Pull request by paras-malhotra**
Laravel ships many routing middleware classes out of the box, but there are no job middleware classes yet. I think it may be very useful if we also introduce job middleware out of the box for common use cases.

One common use case is to avoid overlapping jobs. An example use case could be to avoid processing a double refund on an order.

This PR introduces a `PreventOverlappingJobs` middleware that does just that (prevents overlapping jobs). For the refund processing job example above, we could use this middleware like so:

```php
public function middleware()
{
    return [
        new PreventOverlappingJobs($this->order->id)
    ];
}
```

If there is a need to avoid overlapping jobs globally (instead of say by order), we could use it as mentioned below. An example use case could be say updating the search index when a searchable model is updated.

```php
public function middleware()
{
    return [new PreventOverlappingJobs];
}
```

### CODE REVIEW

File: src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+     * @param  mixed  $job
+     * @param  callable  $next
+     * @return mixed
+     */
+    public function handle($job, $next)
+    {
+        $lock = app(Cache::class)->lock($this->getLockKey($job));
+
+        if ($lock->get()) {
+            $next($job);
```
Comment (taylorotwell) [inline src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php:50]: What if the job throws an exception? The lock will never be released.

**Comment by paras-malhotra** [comment]
@taylorotwell thanks for your review. I've fixed exception handling and also added a test to confirm that:
1. Locks are released on job exceptions
2. Exceptions are passed through so that the queue worker can handle them

**Comment by paras-malhotra** [comment]
Also added an option to expire the lock to handle queue worker crashes (if needed).

**Comment by taylorotwell** [comment]
So if the lock exists... the job is just deleted? Am I missing something? That doesn't seem like desirable behavior.

**Comment by paras-malhotra** [comment]
If the lock exists, the job is "skipped" as it is considered to a duplicate concurrent job. This is similar to the behaviour of a scheduled cron job when run with `withoutOverlapping`. Some use cases would be initiating a refund on the same order twice while the refund is processing, request to update the search index while the indexing is taking place, etc.

I do understand that there can be more use cases where users might want the job released back to the queue. So, I think I'll add an option to do so to cover more use cases.

**Comment by paras-malhotra** [comment]
@taylorotwell I've updated the middleware so that the default behaviour is to release the job back to the queue with an option to not release (or delete on overlap).

This code releases back to the queue on overlap:
```php
public function middleware()
{
    return [
        new PreventOverlappingJobs($this->order->id)
    ];
}
```

This code doesn't release (deletes) on overlap:
```php
public function middleware()
{
    return [
        (new PreventOverlappingJobs($this->order->id))->dontRelease()
    ];
}
```

This code releases after a 60 seconds delay:
```php
public function middleware()
{
    return [
        (new PreventOverlappingJobs($this->order->id))->releaseAfter(60)
    ];
}
```

You could also automatically expire the lock to take care of queue worker crashes:
```php
public function middleware()
{
    return [
        (new PreventOverlappingJobs($this->order->id))->releaseAfter(60)->expireAt(3600)
    ];
}
```

**Comment by themsaid** [comment]
I think this middleware can be more useful if you can configure the maximum concurrency.

Something like:

```php
public function middleware()
{
    return [
        new LimitConcurrency($this->user->id, 5)
    ];
}
```

This will prevent a certain user from having this job running more than 5 instances at the same time.

**Comment by taylorotwell** [comment]
@themsaid that could probably be a separate middleware.

**Comment by paras-malhotra** [comment]
Thanks @themsaid and @taylorotwell, let me know if you'd like me to make any changes.

File: src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+     *
+     * @var int|null
+     */
+    public $releaseAfter;
+
+    /**
+     * Create a new overlapping jobs middleware instance.
+     *
+     * @param  string  $key
+     * @param  int|null  $releaseAfter
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php:41]: can this be null?

File: src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+     * @var int
+     */
+    public $expiresAt;
+
+    /**
+     * The key of the job.
+     *
+     * @var string
+     */
+    public $key;
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php:21]: Order of these properties is odd.

File: src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        $this->prefix = $prefix;
+
+        return $this;
+    }
+
+    /**
+     * Get the lock key.
+     *
+     * @param  mixed  $job
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php:129]: object

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php (hunk 2 above)
Comment (paras-malhotra) [reply src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php:41]: Yes, it's null if we don't want to release the job afterwards. See `dontRelease` and `handle`.

File: src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php (hunk 3 above)
Comment (paras-malhotra) [reply src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php:21]: Current order is alphabetic. I thought that's what Laravel uses?

File: src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php (hunk 4 above)
Comment (paras-malhotra) [reply src/Illuminate/Queue/Middleware/PreventOverlappingJobs.php:129]: This is the same docblock as the `$command` in `CallQueuedHandler` as it's the same variable passed through middleware.

**Comment by taylorotwell** [comment]
Renamed to `WithoutOverlapping`.

**Comment by liamkeily** [comment]
I am experimenting with this middleware in a project and when stress testing i noticed that i hit max attempts because each time the job is released back to the queue it counts as another attempt. I can get around this by defining a  retryUntil but just wondering if it is expected behaviour to increment attempts when the job didn't actually fail?

**Pull request merged**